tests/PHPStan/Analyser/data/loose-comparisons.php

```diff
@@ -518,5 +600,881 @@ public function sayEmptyStr(
+	 */
+	public function looseArray(
+		$true,
+		$false,
+		$one,
+		$zero,
+		$minusOne,
+		$oneStr,
+		$zeroStr,
+		$minusOneStr,
+		$null,
+		$emptyArr,
+		$phpStr,
+		$emptyStr,
+		array $arr,
+		int $int,
+		float $float,
+		bool $bool,
+		string $string,
+		object $obj,
+		$unionMaybeNumeric,
+		$unionNumbers,
+		$unionStrings,
+		$unionMaybeArray
+	) {
+		assertType('bool', $arr == $true);
+		assertType('bool', $arr == $false);
+		assertType('false', $arr == $one);
+		assertType('false', $arr == $zero);
+		assertType('false', $arr == 10);
+		assertType('false', $arr == $minusOne);
+		assertType('false', $arr == $oneStr);
+		assertType('false', $arr == $zeroStr);
+		assertType('false', $arr == $minusOneStr);
+		assertType('false', $arr == $null);
+		assertType('bool', $arr == $emptyArr);
+		assertType('false', $arr == $phpStr);
+		assertType('false', $arr == $emptyStr);
+		assertType('false', $arr == $float);
+		assertType('bool', $arr == []);
+		assertType('true', $arr == $arr);
+		assertType('false', $arr == $int);
+		assertType('false', $arr == $float);
+		assertType('bool', $arr == $bool);
+		assertType('false', $arr == $string);
+		assertType('false', $arr == $obj);
+		assertType('false', $arr == new \stdClass());
+		assertType('false', $arr == $unionMaybeNumeric);
+		assertType('false', $arr == $unionNumbers);
+		assertType('false', $arr == $unionStrings);
+		assertType('bool', $arr == $unionMaybeArray);
+	}
+
+	/**
+	 * @param callable $callable
+	 * @param true $true
+	 * @param false $false
+	 * @param 1 $one
+	 * @param 0 $zero
+	 * @param -1 $minusOne
+	 * @param '1' $oneStr
+	 * @param '0' $zeroStr
+	 * @param '-1' $minusOneStr
+	 * @param null $null
+	 * @param array{} $emptyArr
+	 * @param 'php' $phpStr
+	 * @param '' $emptyStr
+	 * @param array $arr
+	 * @param 'a'|'123'|'123.23' $unionMaybeNumeric
+	 * @param 1|2|3 $unionNumbers
+	 * @param 'a'|'b'|'c' $unionStrings
+	 * @param 'a'|'123'|123|array $unionMaybeArray
+	 *
+	 * https://3v4l.org/RHc0P
+	 */
+	public function looseCallable(
+		$callable,
+		$true,
+		$false,
+		$one,
+		$zero,
+		$minusOne,
+		$oneStr,
+		$zeroStr,
+		$minusOneStr,
+		$null,
+		$emptyArr,
+		$phpStr,
+		$emptyStr,
+		array $arr,
+		int $int,
+		float $float,
+		bool $bool,
+		string $string,
+		object $obj,
+		$unionMaybeNumeric,
+		$unionNumbers,
+		$unionStrings,
+		$unionMaybeArray
+	) {
+		assertType('bool', $callable == 'myFunction');
+		assertType('true', $callable == $true);
+		assertType('false', $callable == $false);
+		assertType('true', $callable == $one);
+		assertType('false', $callable == $zero);
+		assertType('false', $callable == 10);
+		assertType('false', $callable == $minusOne);
+		assertType('bool', $callable == $oneStr); // could be false, because invalid function name
+		assertType('bool', $callable == $zeroStr); // could be false, because invalid function name
+		assertType('bool', $callable == $minusOneStr); // could be false, because invalid function name
```

Comment:
within the type system we cannot use reflection to check whether given strings are valid callables.
we could do some kind of regular expression matching, whether a constant string contains only of chars allowed within a function name, but I don't think its worth it